src/NemoStuff.jl

```diff
@@ -0,0 +1,804 @@
+    return size(base_ring(base_ring(R)))^degree(modulus(R))
+end
+
+function size(R::Union{Generic.ResidueRing{T},Generic.ResidueField{T}}) where {T<:PolyRingElem}
+    return size(base_ring(base_ring(R)))^degree(R.modulus)
+end
+
+function rand(R::Union{Generic.ResidueRing{T},Generic.ResidueField{T}}) where {T<:PolyRingElem}
+    r = rand(base_ring(base_ring(R)))
+    g = gen(R)
+    for i = 1:degree(R.modulus)
+        r = r * g + rand(base_ring(base_ring(R)))
+    end
+    return r
+end
+
+function gens(R::Union{Generic.ResidueRing{T},Generic.ResidueField{T}}) where {T<:PolyRingElem} ## probably needs more cases
+    ## as the other residue functions
+    g = gen(R)
+    r = Vector{typeof(g)}()
+    push!(r, one(R))
+    if degree(R.modulus) == 1
+        return r
+    end
+    push!(r, g)
+    for i = 2:degree(R.modulus)-1
+        push!(r, r[end] * g)
+    end
+    return r
+end
+
+AbstractAlgebra.promote_rule(::Type{LocElem{T}}, ::Type{T}) where {T} = LocElem{T}
+
+AbstractAlgebra.promote_rule(::Type{T}, ::Type{S}) where {S<:NumFieldElem,T<:Integer} = S
+
+AbstractAlgebra.promote_rule(::Type{S}, ::Type{T}) where {S<:NumFieldElem,T<:Integer} = S
+
+function mulmod(a::S, b::S, mod::Vector{S}) where {S<:MPolyRingElem{T}} where {T<:RingElem}
+    return Base.divrem(a * b, mod)[2]
+end
+
+Base.:\(f::Map, x) = preimage(f, x)
+
+function preimage(f::AbstractAlgebra.Generic.CompositeMap, a)
+    return preimage(f.map1, preimage(f.map2, a))
+end
+
+export set_precision, set_precision!
+
+function set_precision(f::PolyRingElem{T}, n::Int) where {T<:SeriesElem}
+    g = parent(f)()
+    for i = 0:length(f)
+        setcoeff!(g, i, set_precision(coeff(f, i), n))
+    end
+    return g
+end
+
+function set_precision!(f::PolyRingElem{T}, n::Int) where {T<:SeriesElem}
+    for i = 0:length(f)
+        setcoeff!(f, i, set_precision!(coeff(f, i), n))
+    end
+    return f
+end
+
+function Base.minimum(::typeof(precision), a::Vector{<:SeriesElem})
+    return minimum(map(precision, a))
+end
+
+function Base.maximum(::typeof(precision), a::Vector{<:SeriesElem})
+    return maximum(map(precision, a))
+end
+
+#TODO: in Nemo, rename to setprecision
+#      fix/report series add for different length
+function set_precision(a::SeriesElem, i::Int)
+    b = deepcopy(a)
+    set_precision!(b, i)
+    return b
+end
+
+ngens(R::MPolyRing) = nvars(R)
+
+Random.gentype(::Type{T}) where {T<:FinField} = elem_type(T)
+
+import LinearAlgebra
+LinearAlgebra.dot(a::RingElem, b::RingElem) = a * b
+
+function is_upper_triangular(M::MatElem)
+    n = nrows(M)
+    for i = 2:n
+        for j = 1:min(i - 1, ncols(M))
+            if !iszero(M[i, j])
+                return false
+            end
+        end
+    end
+    return true
+end
+
+transpose!(A::MatrixElem) = transpose(A)
+
+function Base.div(f::PolyRingElem, g::PolyRingElem)
+    q, r = divrem(f, g)
+    return q
+end
+
+function Base.rem(f::PolyRingElem, g::PolyRingElem)
+    return mod(f, g)
```

Comment:
Huh, I wonder why not
```suggestion
    q, r = divrem(f, g)
    return r
```